text/3550-new-range.md

```diff
@@ -0,0 +1,441 @@
+
+```rust
+for x in it.take(3) {  // a *copy* of the iterator is used here
+    // ..
+}
+
+match it.next() {  // the original iterator (not advanced) is used here
+    // ..
+}
+```
+
+[`Range`]: https://doc.rust-lang.org/stable/core/ops/struct.Range.html
+[`RangeFrom`]: https://doc.rust-lang.org/stable/core/ops/struct.RangeFrom.html
+[`RangeInclusive`]: https://doc.rust-lang.org/stable/core/ops/struct.RangeInclusive.html
+
+However, there is considerable demand for `Copy` range types for multiple reasons:
+- ergonomic use without needing explicit `.clone()`s or rewriting the `a..b` syntax repeatedly
+- use in `Copy` types (currently people work around this by using a tuple instead)
+
+Another primary motivation is the extra size of `RangeInclusive`. It uses an extra `bool` field to keep track of when the upper bound has been yielded by the iterator, but this extra size is useless when the type is not used as an iterator.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rust has several different types of "range" syntax, including the following:
+
+- `a..b` denotes a range from `a` (inclusive) to `b` (exclusive). It resolves to the type `std::ops::range::Range`.  
+  The iterator for `Range` will yield values from `a` (inclusive) to `b` (exclusive) in steps of one.
+
+- `a..=b` denotes a range from `a` (inclusive) to `b` (inclusive). It resolve to the type `std::ops::range::RangeInclusive`.
+  The iterator for `RangeInclusive` will yield values from `a` (inclusive) to `b` (inclusive) in steps of one.
+
+- `a..` denotes a range from `a` (inclusive) with no upper bound. It resolves to the type `std::ops::range::RangeFrom`.  
+  The iterator for `RangeFrom` will yield values starting with `a` and increasing in steps of one.
+
+These types implement the `IntoIterator` trait, enabling their use directly in a `for` loop:
+```rust
+for n in 0..5 {
+    // `n` = 0, 1, 2, 3, 4
+}
+```
+
+All range types are `Copy` when the bounds are `Copy`, allowing easy reuse:
+```rust
+let range = 0..5;
+
+if a_slice[range].contains(x) {
+    // ...
+}
+if b_slice[range].contains(y) {
+    // ...
+}
+```
+
+For convenience, several commonly-used methods from `Iterator` are present as inherent functions on the range types:
+```rust
+for n in (1..).map(|x| x * 2) {
+    // n = 2, 4, 6, 8, 10, ...
+}
+for n in (0..5).rev() {
+    // n = 4, 3, 2, 1, 0
+}
+```
+
+## Legacy Range Types
+
+In Rust editions prior to 2024, `a..b`, `a..=b`, and `a..` resolved to a different set of types (now found in `std::ops::range::legacy`). These legacy range types did not implement `Copy`, and implemented `Iterator` directly (rather than `IntoIterator`).
+
+This meant that any `Iterator` method could be called on those range types:
+```rust
+let mut range = 0..5;
+assert_eq!(range.next(), Some(0));
+range.for_each(|n| {
+    // n = 1, 2, 3, 4
+});
+```
+
+There exist `From` impls for converting from the new range types to the legacy range types.
```

Comment:
Again, I strongly disagree that "legacy" is an incorrect framing here, especially with the proposal to name a module explicitly `legacy`. These should be used as the iterators returned by the `ImplIterator` implementation.

Additionally, it would be against convention to separate out the module that the iterator and `IntoIterator` types live in. Personally, I propose this instead:

1. All range types move to a dedicated `std::range` module, outside `std::ops`. These aren't really operators anyway.
2. The types with the old names in `std::range` (like `Range`, `RangeFrom`, etc.) are the new types that implement `IntoIterator`.
3. The current types also exist in this module, but have `Range` replaced with `Iter` (like `Iter`, `IterFrom`, `IterInclusive`, etc).
4. The `Iter` types are re-exported as deprecated aliases in `std::ops` to retain backwards-compatibility.

Note that as part of this, we could also consider moving stuff like `Bound` and `RangeBounds` into the `std::range` module.

But again, as I said, these are not "legacy range types." They're range iterator types, and it just happens to be that on older editions, they'll be the ones that get used if you type a range *instead* of the standard range types. That's all fine.

Reply:
I strongly disagree with using the existing types as the `IntoIter`s for the new types. I tried to address this in [Use legacy range types as the iterators for the new range types](https://github.com/pitaj/rfcs/blob/new-range/text/3550-new-range.md#use-legacy-range-types-as-the-iterators-for-the-new-range-types). Essentially, iterator types conventionally have a very limited API, for good reasons. The existing range types do not fit that mold in the least.

> Additionally, it would be against convention to separate out the module that the iterator and IntoIterator types live in.

This would not be the case if there are separate iterator types.

1. I'm not attached to the (pretty heavily nested) `std::ops::range::` path, but I must disagree with the assertion that they "aren't really operators" - they are definitively binary operators. That said, I'm okay with `std::range::` - I'll add this as an Unresolved question for **T-libs-api** to decide.
2. Seems to be in agreement with the RFC
3. (As already said) I disagree with using the old types like that, but I'd be fine that that naming. I'll add a notice in the RFC that exact names and module paths are for demonstration purposes only, and can be finalized by **T-libs-api**
4. Seems to be in agreement with the RFC, insofar as the aliases must exist for backwards-compatibility.

> Note that as part of this, we could also consider moving stuff like Bound and RangeBounds into the std::range module.

I suggested the same in **Unresolved questions**:

> Should other range-related items (like `RangeBounds`) also be moved under `std::ops::range`?

Reply:
I guess I should have read the whole thing again before commenting-- for what it's worth, I *also* disagree with that approach.

"Much larger API surface" is quite an exaggeration IMHO: `Range`, for example, has two public fields (whose direct access could be deprecated) and two methods: `contains` and `is_empty`. Far from a massive set of methods.

Also, the range types *have* been heavily optimised; that's why `RangeInclusive` has the boolean field, since this was heavily tested before they were stabilised. The biggest concern I can see is people using the `start` and `end` fields directly on these iterators, which… isn't really that big of a deal, to be honest. Like, we're already dealing with a compromise since these types can't be outright removed, so, I don't know why you'd want to create an entirely new set of iterator types when these are going to be effectively just copies of those.

Reply:
> "Much larger API surface" is quite an exaggeration IMHO: Range, for example, has two public fields (whose direct access could be deprecated) and two methods: contains and is_empty. Far from a massive set of methods.

Public fields are a huge deal, I'm not aware of any iterator type outside of the ranges that has a public field. But you can't just count public fields and methods, you also have to consider trait implementations. The range types have the following trait impls that are unrelated to iteration:
- Default
- PartialEq, Eq
- Hash
- Index, IndexMut, SliceIndex
- RangeBounds

As far as I can tell, no iterators in the standard library but the ranges implement any of those traits except `Default`. We can fix that inconsistency.

`RangeInclusive` specifically must take care with equality, hashing, and indexing because it can be exhausted. By removing those impls from the iterator for it, we can prevent that misuse entirely.

There's also the `ExactSizeIterator` issue: range iterators should never have implemented `ExactSizeIterator` for `u32` because it does not always fit within `usize`. Introducing new types allows us to fix that issue.

> Also, the range types have been heavily optimised; that's why `RangeInclusive` has the boolean field, since this was heavily tested before they were stabilised.

`RangeInclusive` in particular is known to perform sub-optimally (compared to `start..end+1`), so it's weird that you specifically bring it up. Heck, we have people in this RFC discussion providing alternative implementations that could bring performance benefits. It's not clear if those implementations could be made compatible with the existing API surface.

Reply:
Even if the API is larger than normal for an Iterator implementation (and I will point out that in addition to the methods @clarfonthey mentioned, `Range` also implements `Index`, `IndexMut`,  `RangeBounds`, and `SliceIndex`), I think that reusing an existing type that has a larger API than the ideal is better than introducing whole new types with equivalent functionality whose only purpose is to have a smaller API. 

The point about optimization might be more compelling _if_ it can be shown that reducing the API somehow unlocks optimizations that wouldn't otherwise be possible.

Reply:
@tmccombs I'm a little confused because it looks like you responded after me but didn't see my reply

Reply:
> Also, the range types _have_ been heavily optimised; that's why `RangeInclusive` has the boolean field, since this was heavily tested before they were stabilised.

`RangeInclusive` has been heavily optimized _within the constraint of being its own iterator_. It's unlikely it can be further optimized _with this contraint_, but lifting the constraint should allow bringing its performance more in line with that of `Range`, more consistently.

There are numerous reports of performance cliffs hit when using `RangeInclusive` instead of `Range`, here is an artificially constructed test-case for your consideration https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=240d5606fcbfcdf6631aec39be9bb4b9

The short of it is that both `Range`, and an hypothetical optimized RangeInclusive Iterator defined as:

```rust
enum OptInclusiveIter {
    Exclusive(Range<i32>),
    Shifted(Range<i32>),
    Inclusive(RangeInclusive<i32>),
}
```

Both lead to vectorized code, whereas the current `RangeInclusive` implementation does not.

As rustc and LLVM change over time, which test cases do optimize correctly (or not) also change over time, so reusing old test cases doesn't always reproduce the issue -- or not as dramatically -- but there are always test cases where the difference is... drastic. As summing `i * (i / 2)` over `1..=max` (or `1..(max + 1)`) demonstrates here.

Reply:
You are correct, I did not see your reply before I responded. 

> There's also the ExactSizeIterator issue (maybe the strongest argument)

But this isn't even currently mentioned as a reason that this alternative was rejected in the RFC itself.

> RangeInclusive in particular is known to perform sub-optimally (compared to start..end+1)

Again, I think that the RFC could benefit from a little more explanation about this. Specifically pointing out the the current API constrains possible optimizations for `RangeInclusive` (which if I am understanding correctly, is because of the public fields). This also seems like it is specific to `RangeInclusive`. So other than consistency, isn't really an argument for the other Range types.

Reply:
> Again, I think that the RFC could benefit from a little more explanation

Yeah, I am working on a revision to clarify wording and expand on the concerns that have been brought up.

Reply:
> The short of it is that both Range, and an hypothetical optimized RangeInclusive Iterator defined as:

If you use that enum for Range then wouldn't you cause code bloat?  You could make RangeInclusive reuse the simpler Range impl though.